### PR TITLE
I26 use urls

### PR DESCRIPTION
**Pull request by eyvorchuk**
This PR resolves [issue 26](https://github.com/pacificclimate/osprey/issues/26), which is to allow the config file for `wps_parameters.py` to have `FILE_PATH`s pointing to data files on THREDDS. For each URL, the data was copied to a temporary file in the working directory of the process, and the URL in the config file was replaced with the corresponding local file path. This functionality was tested by mocking the URLs prior to running the process to avoid requesting data that may not exist on THREDDS.

### CODE REVIEW

**Comment by eyvorchuk** [comment]
@jameshiebert I should mention that some of the `RVIC` packages in the `venv` need to be modified before the tests will work. The necessary changes are [in the testing section of the README](https://github.com/pacificclimate/osprey#testing).

File: tests/configs/parameter_https.cfg [hunk 1]
```diff
@@ -73,7 +73,7 @@ SEARCH_FOR_CHANNEL: False
 #-- Path to Pour Points File (char) --#
 # A comma separated file of outlets to route to [lons, lats] - one coordinate pair per line (order not important)
 # May optionally include a column [names] - which will (if not aggregating) be included in param file
-FILE_NAME: https://docker-dev03.pcic.uvic.ca/twitcher/ows/proxy/thredds/dodsC/datasets/RVIC/sample_pour.txt
+FILE_NAME: https://docker-dev03.pcic.uvic.ca/twitcher/ows/proxy/thredds/fileServer/datasets/RVIC/sample_pour.txt
```
Comment (sum1lim) [inline tests/configs/parameter_https.cfg:77]: I think the ideal use of `pour` is to get user inputs (lon and lat) and create a `pour.txt` file in the `tmp` directory. It is a variable that needs to be dynamically defined by users. I guess we don't have to worry about that here though. I will write an issue.

File: tests/test_wps_parameters.py [hunk 2]
```diff
@@ -21,3 +21,19 @@ def test_parameters_local(config):
         temp_config.read()
         params = f"config={temp_config.name};"
         run_wps_process(Parameters(), params)
+
+
+@pytest.mark.online
+@pytest.mark.parametrize(
+    ("config"), [resource_filename(__name__, "configs/parameter_https.cfg")],
+)
+def test_parameters_https(config, make_mock_urls):
```
[the remark this answers is not in the document]
Comment (eyvorchuk) [reply tests/test_wps_parameters.py:30]: What does each `FILE_PATH` in your `parameter_https.cfg` file look like? I just realized that running the notebook as is permanently changes it.

[the remark this answers is not in the document]
Comment (eyvorchuk) [reply tests/test_wps_parameters.py:30]: Actually, the demo doesn't use that `.cfg` file, so it won't be modified. I'll look into it more.

[the remark this answers is not in the document]
Comment (eyvorchuk) [reply tests/test_wps_parameters.py:30]: Could you try [replacing self.workdir in this line with a different directory](https://github.com/pacificclimate/osprey/blob/4283862881ddc828fa4f43acdb47342ce73a6ae7/osprey/processes/wps_parameters.py#L101), and [after the for loop in replace_urls, print filedata](https://github.com/pacificclimate/osprey/blob/4283862881ddc828fa4f43acdb47342ce73a6ae7/osprey/utils.py#L64)? I just want to see if the contents of the config file are being modified correctly.

**Comment by jameshiebert** [comment]
> @jameshiebert I should mention that some of the `RVIC` modules in the `venv` need to be modified before the tests will work. The necessary changes are [in the testing section of the README](https://github.com/pacificclimate/osprey#testing).

Could you [fix this](https://github.com/UW-Hydro/RVIC/issues/96#issuecomment-360191320) in RVIC and open a PR against RVIC [as requested by Joe Hamman](https://github.com/UW-Hydro/RVIC/issues/96#issuecomment-360191320) for your \#2 testing note? Testing should be automated; it's not feasible to have to make manual changes to an installed virtual environment. We need to fix the root of this problem. And if its as simple [as you note](https://github.com/pacificclimate/osprey#testing), it should be a no brainer.

**Comment by sum1lim** [comment]
@jameshiebert I already made a [PR](https://github.com/UW-Hydro/RVIC/pull/132) in RVIC repo and emailed Joe Hamman, but still waiting for the reply.

**Comment by eyvorchuk** [comment]
> Could you [fix this](https://github.com/UW-Hydro/RVIC/issues/96#issuecomment-360191320) in RVIC and open a PR against RVIC [as requested by Joe Hamman](https://github.com/UW-Hydro/RVIC/issues/96#issuecomment-360191320) for your second testing note? Testing should be automated; it's not feasible/scalable to have to make manual changes to an installed virtual environment. We need to fix the root of this problem. And if its as simple [as you note](https://github.com/pacificclimate/osprey#testing), it should be a no brainer.

Unfortunately, we haven't had luck with getting those issues fixed in RVIC. [The first issue](https://github.com/UW-Hydro/RVIC/issues/96) has been fixed in [RVIC 1.1.1](https://github.com/UW-Hydro/RVIC/blob/5987da6d33cea8063c848b60215353915ac2bb10/rvic/core/share.py#L277), but for some reason, that version isn't installable as a python package (the only one available is v1.1.0.post1). @sum1lim [opened a PR for the other issue a month ago](https://github.com/UW-Hydro/RVIC/pull/132) and I believe emailed one of the devs about it, but neither of those have gotten responses.

File: tests/test_wps_parameters.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sum1lim) [reply tests/test_wps_parameters.py:30]: > What does each FILE_PATH in your parameter_https.cfg file look like? I just realized that running the notebook as is permanently changes it.

I think you are right here. I did `git stash` and ran pytest again, and it passed. Last time I ran the demo first, so that caused the issue you said.

File: osprey/cli.py [hunk 3]
```diff
@@ -152,7 +152,7 @@ def stop():
 @click.option(
     "--parallelprocesses",
     metavar="INT",
-    default="2",
+    default="20",
```
Comment (nikola-rados) [inline osprey/cli.py:155]: Revert this change

File: osprey/utils.py [hunk 4]
```diff
@@ -34,6 +36,38 @@ def replace_filenames(config, temp_config):
...
+    """
+    Copy https URLs to local storage and replace URLs
+    with local paths in config file.
+    Parameters:
+        config (str): Config file
+        outdir (str): Output directory
+    """
+    read_config = open(config, "r")
+    filedata = read_config.readlines()
+    read_config.close()
```
Comment (nikola-rados) [inline osprey/utils.py:49]: We could use a context manager here.

File: osprey/utils.py [hunk 5]
```diff
@@ -34,6 +36,38 @@ def replace_filenames(config, temp_config):
...
+            url = url.rstrip()  # remove \n character at end
+            r = requests.get(url)
+            filename = url.split("/")[-1]
+            prefix, suffix = filename.split(".")
+            suffix = "." + suffix
+            local_file = NamedTemporaryFile(
+                suffix=suffix, prefix=prefix, dir=outdir, delete=False
+            )
+            local_file.write(r.content)
+            filedata[i] = filedata[i].replace(url, local_file.name)
```
Comment (nikola-rados) [inline osprey/utils.py:63]: While this is _fine_, we generally don't want to use indices if we can help it. It is more understandable to iterate through the iterable object : `for data in filedata:...`.

File: osprey/utils.py [hunk 6]
```diff
@@ -34,6 +36,38 @@ def replace_filenames(config, temp_config):
...
+        if "https" in filedata[i]:
+            url = filedata[i].split(" ")[-1]  # https url is last word in line
+            url = url.rstrip()  # remove \n character at end
+            r = requests.get(url)
+            filename = url.split("/")[-1]
+            prefix, suffix = filename.split(".")
+            suffix = "." + suffix
+            local_file = NamedTemporaryFile(
+                suffix=suffix, prefix=prefix, dir=outdir, delete=False
+            )
```
Comment (nikola-rados) [inline osprey/utils.py:61]: Great use of tempfile here :+1:

File: osprey/utils.py [hunk 7]
```diff
@@ -34,6 +36,38 @@ def replace_filenames(config, temp_config):
...
+            local_file.write(r.content)
+            filedata[i] = filedata[i].replace(url, local_file.name)
+
+    write_config = open(config, "w")
+    for line in filedata:
+        write_config.write(f"{line}")
+    write_config.close()
```
Comment (nikola-rados) [inline osprey/utils.py:68]: Again, we can use a context manager.

File: requirements.txt [hunk 8]
```diff
@@ -2,6 +2,6 @@ pywps>=4.2
 jinja2
 click
 psutil
-rvic==1.1.0post1
+rvic==1.1.1
```
Comment (nikola-rados) [inline requirements.txt:5]: :+1:

File: requirements_dev.txt [hunk 9]
```diff
@@ -1,6 +1,7 @@
 pytest
 flake8
 pytest-flake8
+requests-mock
```
Comment (nikola-rados) [inline requirements_dev.txt:4]: I didn't know this package existed. I wonder if we can apply to our notebooks to make truly offline tests.

File: tests/configs/parameter_https.cfg [hunk 10]
```diff
@@ -73,7 +73,7 @@ SEARCH_FOR_CHANNEL: False
...
```
Comment (nikola-rados) [inline tests/configs/parameter_https.cfg:76]: What is the purpose of this change from `dodsC` to `fileServer`?

File: tests/conftest.py [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import pytest
+from osprey.testing import make_mock_urls
+
+
+@pytest.fixture
+def conftest_make_mock_urls(config, requests_mock):
+    return make_mock_urls(config, requests_mock)
```
Comment (nikola-rados) [inline tests/conftest.py:7]: Nice fixture!

File: tests/test_wps_parameters.py [hunk 12]
```diff
@@ -21,3 +21,19 @@ def test_parameters_local(config):
...
+    ("config"), [resource_filename(__name__, "configs/parameter_https.cfg")],
+)
+def test_parameters_https(config, conftest_make_mock_urls):
+    config_name = os.path.splitext(config)[0]  # Remove .cfg extension
+    with NamedTemporaryFile(
+        suffix=".cfg", prefix=os.path.basename(config_name), mode="w+t"
+    ) as temp_config:  # Avoid permanent replacement of https URLs
+        read_config = open(config, "r")
+        temp_config.writelines(read_config.read())
+        temp_config.read()
```
Comment (nikola-rados) [inline tests/test_wps_parameters.py:37]: Is there a reason we aren't just directly reading into the `temp_config`?

**Comment by eyvorchuk** [comment]
>  That being said I am getting failures on my `pytest` runs locally.

What do the `FILE_NAME`s in your `.cfg` files look like?

**Comment by nikola-rados** [comment]
`FILE_NAME: https://docker-dev03.pcic.uvic.ca/twitcher/ows/proxy/thredds/fileServer/datasets/RVIC/sample_flow_parameters.nc`

**Comment by eyvorchuk** [comment]
And in `sample_parameter_config.cfg`?

**Comment by nikola-rados** [comment]
`FILE_NAME: tests/data/samples/sample_pour.txt`

And here are the failures I got:
```
==================================================================================================== FAILURES =====================================================================================================
____________________________________________________________ test_wps_convolution[/home/nrados/github/pcic/osprey/tests/configs/convolve_opendap.cfg] _____________________________________________________________
Traceback (most recent call last):
  File "/home/nrados/github/pcic/osprey/tests/test_wps_convolution.py", line 38, in test_wps_convolution
    run_wps_process(Convolution(), params)
  File "/home/nrados/github/pcic/osprey/venv/lib/python3.6/site-packages/wps_tools/testing.py", line 65, in run_wps_process
    assert_response_success(resp)
  File "/home/nrados/github/pcic/osprey/venv/lib/python3.6/site-packages/pywps/tests.py", line 145, in assert_response_success
    assert len(success) == 1
AssertionError
---------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------
curl error details: 
__________________________________________________________________________________________ test_wps_convolution[config1] __________________________________________________________________________________________
Traceback (most recent call last):
  File "/home/nrados/github/pcic/osprey/tests/test_wps_convolution.py", line 38, in test_wps_convolution
    run_wps_process(Convolution(), params)
  File "/home/nrados/github/pcic/osprey/venv/lib/python3.6/site-packages/wps_tools/testing.py", line 65, in run_wps_process
    assert_response_success(resp)
  File "/home/nrados/github/pcic/osprey/venv/lib/python3.6/site-packages/pywps/tests.py", line 145, in assert_response_success
    assert len(success) == 1
AssertionError
---------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------
curl error details: 
```

**Comment by eyvorchuk** [comment]
Oh, convolution is @sum1lim's process. I didn't modify anything involved with it in the PR, so the issues might be happening elsewhere.

**Comment by sum1lim** [comment]
@nikola-rados if only `convolution` process is not working, the possibility is that `valid_range`s are not changed in `share.py`. Could you double-check it?

**Comment by sum1lim** [comment]
OK, I know the problem. @eyvorchuk, could you change this [line](https://github.com/pacificclimate/osprey/blob/i26-use-urls/osprey/processes/wps_convolution.py#L71) to following?
```
"DATL_LIQ_FLDS": ["RUNOFF", "BASEFLOW"],
```
I actually encountered this problem after RVIC 1.1.1 is used.

**Comment by eyvorchuk** [comment]
Works now.

File: osprey/utils.py (hunk 5 above)
Comment (eyvorchuk) [reply osprey/utils.py:63]: I used the index because I wasn't sure if the assignment in line 63 would modify the `filedata` list if I iterated through each element.

File: tests/configs/parameter_https.cfg (hunk 10 above)
Comment (eyvorchuk) [reply tests/configs/parameter_https.cfg:76]: This is so that `https` instead of `OpenDAP` is used to download the file from THREDDS. James told me that it makes downloading entire files easier since I won't have to specify attributes or time ranges.

File: tests/test_wps_parameters.py (hunk 12 above)
Comment (eyvorchuk) [reply tests/test_wps_parameters.py:37]: I'm not sure what you mean here.
